File: public/java/035-lowest-common-ancestor.java
/*
 * LeetCode Problem #236: Lowest Common Ancestor of a Binary Tree
 * Difficulty: Medium
 * Link: https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-tree/
 */
import java.util.*;

public class Solution {

    static class TreeNode {
        int val;
        TreeNode left, right;
        TreeNode(int val) { this.val = val; }
    }

    // APPROACH 1: Recursive | O(n) time | O(h) space
    // EXPLAIN: If root is null/p/q return root; otherwise recurse both sides — if both sides return non-null, root is the LCA; otherwise propagate the non-null result.
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null || root == p || root == q) return root;
        TreeNode left  = lowestCommonAncestor(root.left,  p, q);
        TreeNode right = lowestCommonAncestor(root.right, p, q);
        if (left != null && right != null) return root;
        return left != null ? left : right;
    }

    // APPROACH 2: Parent Pointers | O(n) time | O(n) space
    // EXPLAIN: BFS/DFS to build a parent-pointer map; trace p's ancestors into a set, then walk q's ancestors until finding one in the set.
    public TreeNode lowestCommonAncestorParents(TreeNode root, TreeNode p, TreeNode q) {
        Map<TreeNode, TreeNode> parent = new HashMap<>();
        Deque<TreeNode> stack = new ArrayDeque<>();
        parent.put(root, null);
        stack.push(root);
        while (!parent.containsKey(p) || !parent.containsKey(q)) {
            TreeNode node = stack.pop();
            if (node.left  != null) { parent.put(node.left,  node); stack.push(node.left);  }
            if (node.right != null) { parent.put(node.right, node); stack.push(node.right); }
        }
        Set<TreeNode> ancestors = new HashSet<>();
        while (p != null) { ancestors.add(p); p = parent.get(p); }
        while (!ancestors.contains(q)) q = parent.get(q);
        return q;
    }

    // APPROACH 3: Path Storage | O(n) time | O(n) space
    // EXPLAIN: Find the full root-to-p and root-to-q paths; the last common node on both paths is the LCA.
    public TreeNode lowestCommonAncestorPaths(TreeNode root, TreeNode p, TreeNode q) {
        List<TreeNode> pathP = new ArrayList<>(), pathQ = new ArrayList<>();
        findPath(root, p, pathP);
        findPath(root, q, pathQ);
        TreeNode lca = null;
        for (int i = 0; i < pathP.size() && i < pathQ.size(); i++) {
            if (pathP.get(i) == pathQ.get(i)) lca = pathP.get(i);
            else break;
        }
        return lca;
    }
    private boolean findPath(TreeNode root, TreeNode target, List<TreeNode> path) {
        if (root == null) return false;
        path.add(root);
        if (root == target) return true;
        if (findPath(root.left, target, path) || findPath(root.right, target, path)) return true;
        path.remove(path.size() - 1);
        return false;
    }
}

// Made with Bob

File: public/python/035-lowest-common-ancestor.py
# LeetCode Problem #236: Lowest Common Ancestor of a Binary Tree
# Difficulty: Medium
# Link: https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-tree/

from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# ─────────────────────────────────────────────
# APPROACH 1: Recursive | O(n) time | O(h) space
# EXPLAIN: Return root if it is None/p/q; if both subtrees find a target, current node is LCA; otherwise propagate the non-None result.
# WHEN: Elegant one-pass solution; standard interview answer.

def lowestCommonAncestor_recursive(root: Optional[TreeNode],
                                    p: TreeNode, q: TreeNode) -> Optional[TreeNode]:
    if not root or root is p or root is q:
        return root
    left  = lowestCommonAncestor_recursive(root.left, p, q)
    right = lowestCommonAncestor_recursive(root.right, p, q)
    if left and right:
        return root
    return left if left else right


# ─────────────────────────────────────────────
# APPROACH 2: Parent Pointers | O(n) time | O(n) space
# EXPLAIN: Build a parent-pointer map with BFS; collect all ancestors of p into a set, then walk q's ancestor chain until a match.
# WHEN: More intuitive for those comfortable with the "two paths from root" mental model.

def lowestCommonAncestor_parents(root: Optional[TreeNode],
                                  p: TreeNode, q: TreeNode) -> Optional[TreeNode]:
    from collections import deque
    parent = {root: None}
    dq = deque([root])
    while p not in parent or q not in parent:
        node = dq.popleft()
        if node.left:
            parent[node.left] = node
            dq.append(node.left)
        if node.right:
            parent[node.right] = node
            dq.append(node.right)
    ancestors = set()
    while p:
        ancestors.add(p)
        p = parent[p]
    while q not in ancestors:
        q = parent[q]
    return q


# ─────────────────────────────────────────────
# APPROACH 3: Path Storage | O(n) time | O(n) space
# EXPLAIN: Find the root-to-p and root-to-q paths; the last node both paths share is the LCA.
# WHEN: Most explicit — useful when you want to inspect the full paths.

def lowestCommonAncestor_paths(root: Optional[TreeNode],
                                p: TreeNode, q: TreeNode) -> Optional[TreeNode]:
    def find_path(node, target, path):
        if not node:
            return False
        path.append(node)
        if node is target:
            return True
        if find_path(node.left, target, path) or find_path(node.right, target, path):
            return True
        path.pop()
        return False

    path_p: list = []
    path_q: list = []
    find_path(root, p, path_p)
    find_path(root, q, path_q)
    lca = None
    for a, b in zip(path_p, path_q):
        if a is b:
            lca = a
        else:
            break
    return lca


# Made with Bob
